geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/ReplayKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,419 @@
+      if (key.ts.isEqual(startTime)) 0 else if (key.ts.isAfter(startTime)) 1 else -1
+    })
+
+    // required: there is only 1 partition;  validate??
+    val stream = kafkaConsumer.createMessageStreams(1, offsetRequest).head
+
+    val msgs = stream.iterator
+      .stopOnTimeout
+      .map(msgDecoder.decode)
+      .dropWhile(replayConfig.isBeforeRealStart)
+      .takeWhile(replayConfig.isNotAfterEnd)
+      .foldLeft(List.empty[GeoMessage])((seq, elem) => elem :: seq)
+      .toArray
+
+    val readTime = (System.currentTimeMillis() - readStart).asInstanceOf[AnyRef]
+    logger.debug("Read {} messages in {}ms from {} using {}",
+      msgs.length.asInstanceOf[AnyRef], readTime, topic, replayConfig)
+
+    msgs
+  }
+}
+
+object ReplayTimeHelper {
+
+  private[kafka] val ff = CommonFactoryFinder.getFilterFactory2
+
+  val AttributeName: String = "KafkaLogTime"
+  val AttributeProp: PropertyName = ff.property(AttributeName)
+
+  def addReplayTimeAttribute(builder: SimpleFeatureTypeBuilder): Unit =
+    builder.add(AttributeName, classOf[Date])
+  
+  def toFilter(time: Instant): Filter =
+    ff.equals(AttributeProp, ff.literal(time.toDate))
+
+  def fromFilter(filter: PropertyIsEqualTo): Option[Long] = {
+    checkOrder(filter.getExpression1, filter.getExpression2)
+      .filter(pl => pl.name == AttributeName && pl.literal.getValue.isInstanceOf[Date])
+      .map(pl => new Instant(pl.literal.getValue.asInstanceOf[Date]).getMillis)
+  }
+}
+
+/** @param sft the [[SimpleFeatureType]] - must contain the replay time attribute
+  * @param replayTime the current replay time
+  */
+class ReplayTimeHelper(sft: SimpleFeatureType, replayTime: Long) {
+  import ReplayTimeHelper._
+
+  private val replayDate = new java.util.Date(replayTime)
+  private val builder = new SimpleFeatureBuilder(sft)
+  private val attrIndex = sft.indexOf(AttributeName)
+
+  require(attrIndex >= 0, s"Invalid SFT.  The $AttributeName attribute is missing.")
+
+  /** Copy the given ``sf`` and add a value for the replay time attribute. */
+  def addReplayTime(sf: SimpleFeature): SimpleFeature = {
+    builder.init(sf)
+    builder.set(attrIndex, replayDate)
+    builder.buildFeature(sf.getID)
+  }
+}
+
+
+/** Represents the state at a specific point in time.
+  *
+  * @param sft the SFT
+  * @param events must be ordered, with the most recent first; must consist of only [[CreateOrUpdate]] and
+  *               [[Delete]] messages
+  */
+private[kafka] case class ReplaySnapshotFeatureCache(override val sft: SimpleFeatureType,
+                                                     replayTime: Long,
+                                                     events: Seq[GeoMessage])
+
+  extends KafkaConsumerFeatureCache {
+
+  override lazy val (qt, features) = processMessages
+
+  private def processMessages: (Quadtree, mutable.Map[String, FeatureHolder]) = {
+    val features = new mutable.HashMap[String, FeatureHolder]()
+    val qt = new Quadtree
```

Comment:
No, each query to a replay consumer feature source results in the creation of a new ReplaySnapshotFeatureCache used only to answer that single query and then its thrown away.